File: C++/2-Medium/00237._删除链表中的节点/00237._删除链表中的节点.cpp
#include <iostream>
using namespace std;

// 定义链表节点结构体
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

// 实现删除节点的函数
void deleteNode(ListNode* node) {
    if (node == nullptr || node->next == nullptr) return; // 如果节点为空或者是尾节点，则不做任何操作
    node->val = node->next->val; // 将下一个节点的值复制到当前节点
    ListNode* toDelete = node->next; // 临时保存下一个节点，以便删除
    node->next = node->next->next; // 将当前节点的next指针指向下一个节点的next
    delete toDelete; // 删除下一个节点
}

// 主函数，用于测试
int main() {
    // 创建链表 4 -> 5 -> 1 -> 9
    ListNode* head = new ListNode(4);
    head->next = new ListNode(5);
    head->next->next = new ListNode(1);
    head->next->next->next = new ListNode(9);
    
    cout << "Original list: ";
    for(ListNode* cur = head; cur != nullptr; cur = cur->next) {
        cout << cur->val << " ";
    }
    cout << endl;
    
    // 删除节点值为5的节点
    deleteNode(head->next); // 假设这是我们要删除的节点
    
    cout << "List after deletion: ";
    for(ListNode* cur = head; cur != nullptr; cur = cur->next) {
        cout << cur->val << " ";
    }
    cout << endl;
    
    // 清理分配的内存
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }

    return 0;
}

File: Python/2-Medium/00237._删除链表中的节点/00237._删除链表中的节点.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def deleteNode(node: ListNode) -> None:
    """
    删除链表中的节点，不返回任何值。假设给定的节点不是链表的尾节点。

    Args:
    node (ListNode): 要删除的节点。
    
    注：由于不可以访问链表的头节点和前一个节点，此函数通过复制下一个节点的值到当前节点，然后删除下一个节点的方式来间接删除当前节点。
    """
    # 将下一个节点的值复制到当前节点
    node.val = node.next.val
    # 将当前节点的next指针指向下下个节点，从而删除下一个节点
    node.next = node.next.next

# 用于打印链表的函数，方便验证结果
def printLinkedList(head: Optional[ListNode]) -> None:
    """
    打印链表。

    Args:
    head (ListNode): 链表的头节点。
    """
    current = head
    while current:
        print(current.val, end=" -> " if current.next else "")
        current = current.next
    print()

# 创建示例链表 4->5->1->9
head = ListNode(4)
head.next = ListNode(5)
node_to_delete = head.next  # 要删除的节点，值为5
head.next.next = ListNode(1)
head.next.next.next = ListNode(9)

# 打印原始链表
print("原始链表:")
printLinkedList(head)

# 删除节点
deleteNode(node_to_delete)

# 打印删除节点后的链表
print("删除节点后的链表:")
printLinkedList(head)